src/transformation/visitors/class/decorators.ts

```diff
@@ -2,42 +2,38 @@ import * as ts from "typescript";
 import * as lua from "../../../LuaAST";
 import { TransformationContext } from "../../context";
 import { decoratorInvalidContext } from "../../utils/diagnostics";
-import { addExportToIdentifier } from "../../utils/export";
 import { ContextType, getFunctionContextType } from "../../utils/function-context";
 import { LuaLibFeature, transformLuaLibFunction } from "../../utils/lualib";
-import { transformIdentifier } from "../identifier";
 
-export function createConstructorDecorationStatement(
-    context: TransformationContext,
-    declaration: ts.ClassLikeDeclaration
-): lua.AssignmentStatement | undefined {
-    const className =
-        declaration.name !== undefined
-            ? addExportToIdentifier(context, transformIdentifier(context, declaration.name))
-            : lua.createAnonymousIdentifier();
-
-    const decorators = declaration.decorators;
-    if (!decorators) {
-        return undefined;
+export function transformDecoratorExpression(context: TransformationContext, decorator: ts.Decorator): lua.Expression {
+    const expression = decorator.expression;
+    const type = context.checker.getTypeAtLocation(expression);
+    const callContext = getFunctionContextType(context, type);
+    if (callContext === ContextType.Void) {
+        context.diagnostics.push(decoratorInvalidContext(decorator));
     }
 
-    const decoratorExpressions = decorators.map(decorator => {
-        const expression = decorator.expression;
-        const type = context.checker.getTypeAtLocation(expression);
-        const callContext = getFunctionContextType(context, type);
-        if (callContext === ContextType.Void) {
-            context.diagnostics.push(decoratorInvalidContext(decorator));
-        }
+    return context.transformExpression(expression);
+}
 
-        return context.transformExpression(expression);
-    });
+export function createDecoratingExpression(
+    context: TransformationContext,
+    decorators: lua.Expression[],
+    targetTableName: lua.Expression,
+    targetFieldExpression?: lua.Expression
+): lua.Expression {
+    const decoratorTable = lua.createTableExpression(decorators.map(e => lua.createTableFieldExpression(e)));
 
-    const decoratorTable = lua.createTableExpression(
-        decoratorExpressions.map(expression => lua.createTableFieldExpression(expression))
-    );
+    if (targetFieldExpression) {
+        return transformLuaLibFunction(
+            context,
+            LuaLibFeature.Decorate,
+            undefined,
+            decoratorTable,
+            targetTableName,
+            targetFieldExpression
+        );
+    }
 
-    return lua.createAssignmentStatement(
-        className,
-        transformLuaLibFunction(context, LuaLibFeature.Decorate, undefined, decoratorTable, className)
-    );
+    return transformLuaLibFunction(context, LuaLibFeature.Decorate, undefined, decoratorTable, targetTableName);
```

Comment:
This looks a bit better to me than that `if (targetFieldExpression)` condition above
```suggestion
    const parameters = [decoratorTable, targetTableName];
    if (targetFieldExpression) parameters.push(targetFieldExpression);
    return transformLuaLibFunction(context, LuaLibFeature.Decorate, undefined, ...parameters);
```